feat: Prefix-based routing in IBCv2 Router

kulikthebird opened the pull request:
## Description

This PR introduces a prefix-based routing for the IBCv2 Router. Before this change only the exact port IDs could be associated with a given IBCModule. This change make it possible to associate port ID prefixes with the modules.

closes: #8302


Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against the correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#pull-request-targeting)).
- [ ] Linked to GitHub issue with discussion and accepted design, OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://docs.cosmos.network/main/build/building-modules/intro) and [Go style guide](../docs/dev/go-style-guide.md).
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/ibc-go/blob/main/testing/README.md#ibc-testing-package).
- [ ] Updated relevant documentation (`docs/`).
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Provide a [conventional commit message](https://github.com/cosmos/ibc-go/blob/main/docs/dev/pull-requests.md#commit-messages) to follow the repository standards.
- [ ] Include a descriptive changelog entry when appropriate. This may be left to the discretion of the PR reviewers. (e.g. chores should be omitted from changelog)
- [ ] Re-reviewed `Files changed` in the GitHub PR explorer.
- [ ] Review `SonarCloud Report` in the comment section below once CI passes.

AdityaSripal reviewed (commented):
Should we have some way to ensure that there is no portID that is a prefix of another portID?

kulikthebird commented:
> Should we have some way to ensure that there is no portID that is a prefix of another portID?

Good point - I've added a check for that and updated the interface a little bit.

AdityaSripal reviewed (commented):
lgtm so far, will discuss internally if we want to use the new dependency or just implement prefix check directly as we did with the IBC v1 router

AdityaSripal commented:
Hi @kulikthebird can you please remove this external dependency and just use a for-loop as we had originally. We decided internally not to add the additional third-party library for now.

Thanks! I will approve and merge once that change is in

AdityaSripal reviewed (approved):
Thank you @kulikthebird !!

gjermundgaraba commented on modules/core/api/router.go:
```diff
@@ -20,32 +21,44 @@
+}
+
+func (rtr *Router) getRoute(portID string) (string, IBCModule, bool) {
+	for prefix, module := range rtr.routes {
+		if strings.HasPrefix(portID, prefix) {
+			return prefix, module, true
+		}
+	}
```
Is there any scenario where this could actually cause any problems?
@AdityaSripal?

AdityaSripal replied:
The issue here as I understand is that order is not deterministic here. So the first match will pass. But this is why we ensure that no two portIDs on the router are prefixes of each other. Thus the first portID to match the prefix check **should** be the only portID to match the prefix check.

cc: @kulikthebird

kulikthebird replied:
I've just realized, that it is possible to add a shorter prefix using `AddRoute` method i.e.
```
1. Add prefix "port01"
2. Try to add prefix "port01somesuffix" - this won't pass, because we have a prefix "port01"
3. Add prefix "port" - this would pass, because non of the already added prefixes is a prefix for "port"
```

I'll add an additional check and test case for that scenario

gjermundgaraba reviewed (commented):
Seesm reasonable to me. I want to make sure we are confident that the map iteration is safe though. Maybe even put a comment on why it is safe, and under what types of usage it might not (to prevent someone from accidentally using it for something else)

gjermundgaraba reviewed (approved):
This looks good! One last ask, could you please add a changelog entry as well? 🙏
